### ISSUE TITLE
[codegen] arith.truncf f32->f16 followed by arith.extf f16->f32 is folded to identity by SPIR-V/IGC, breaking numerical barrier

### ISSUE THREAD

**Issue by ZhaoqiongZ** [intel/intel-xpu-backend-for-triton#7491]
## Describe the bug

`arith.truncf f32 -> f16` in the XPU backend is lowered to a plain
unconstrained `LLVM::FPTruncOp`. When the truncation is immediately
followed by `arith.extf f16 -> f32` (a numerical barrier used by, e.g.,
PyTorch Inductor's `emulate_precision_casts`), the downstream SPIR-V /
IGC lowering folds the `fptrunc`+`fpext` pair away as identity, so the
fp16 rounding never occurs at runtime. The same pattern lowers correctly
on CUDA (nvptx keeps the round-trip). CUDA parity is broken.

### Minimal reproducer (hand-written Triton kernel, no PyTorch)

```python
import torch
import triton
import triton.language as tl

@triton.jit
def roundtrip_kernel(X, Y, N: tl.constexpr):
    off = tl.arange(0, N)
    x = tl.load(X + off)
    y = x.to(tl.float16).to(tl.float32)   # explicit fp16 round-trip
    tl.store(Y + off, y)

# A value whose nearest fp16 is different from its fp32 representation.
x = torch.tensor([-2.643845], device="xpu", dtype=torch.float32)
y = torch.empty_like(x)
roundtrip_kernel[(1,)](x, y, 1)

print("input :", x.item())         # -2.643845
print("output:", y.item())         # observed on XPU: -2.643845 (round-trip elided)
print("fp16  :", x.to(torch.float16).float().item())  # -2.64453125 (expected)
```

Expected: `output == -2.64453125` (bit-exact fp16 rounding).
Actual on XPU: `output == -2.643845` — the fp16 narrowing is elided.

### Effect on PyTorch (downstream)

The upstream PyTorch UT
`test/inductor/test_cuda_repro.py::CudaReproTests::test_emulate_precision_casts_preserves_explicit_precision_cast_float16`
fails on XPU with a max abs diff of ~0.018 for a small `matmul -> .to(fp16).float() -> sigmoid -> sum` graph
(`torch._inductor.config.emulate_precision_casts = True`). Isolation
narrows the entire error to the `.to(fp16).float()` step alone
(diff ~7.7e-3 with no reduce), matching the standalone Triton reproducer
above.

Currently disabled and tracked at:
- pytorch/pytorch#190157 (DISABLED marker)
- pytorch/pytorch#190163 (source-level skip PR)
- intel/torch-xpu-ops#4358 (canonical tracker)

### Diagnosis (what I've verified locally)

- Inductor codegen is correct: the emitted Triton source contains
  `.to(tl.float16).to(tl.float32)`.
- TTIR / TTGIR / LLIR (dumped via `MLIR_ENABLE_DUMP=1 TRITON_KERNEL_DUMP=1`)
  all retain the pair. LLIR shows:
  ```
  %22 = fptrunc float %21 to half
  %23 = fpext   half  %22 to float
  ```
  So the folding happens **after** LLIR (SPIR-V translation or IGC).
- Setting `TRITON_INTEL_DISABLE_IGC_OPT=1` (which passes `-cl-opt-disable`
  to ocloc) does **not** change the numerical result, so the folding is
  not gated by IGC's user-tunable optimization level.
- `TORCHINDUCTOR_USE_FAST_MATH` / `TRITON_INTEL_FAST_MATH` are not set;
  `set_fast_math` runs with `enableFpFusion=False` (Inductor sets it to
  `not config.emulate_precision_casts`), so no fast-math flags are added
  to the LLVM instructions.

### Suspected root cause

In `third_party/intel/lib/TritonIntelGPUToLLVM/ElementwiseOpToLLVM.cpp`,
`TruncFOpConversion::createDestOps` (around line 1442):

```cpp
return {LLVM::FPTruncOp::create(rewriter, loc, elemTy, operands[0][0])};
```

This is an unconstrained `fptrunc`. Compare with the sibling
`FpToFpOpConversion::convertFp32ToFp16` (around line 1097), which uses
`LLVM::ConstrainedFPTruncIntr` (constrained intrinsic with explicit
rounding mode and fp exception behavior). Constrained intrinsics carry
side-effect metadata that prevents the downstream folding; the
`tt.fp_to_fp` path is not affected by this bug for that reason.

The two lowering paths for the same source-level operation
(`x.to(tl.float16)`) end up in different arms depending on whether the
op is materialized as `tt.fp_to_fp` or `arith.truncf` after upstream
canonicalization, and they should be numerically equivalent.

### Suggested fix (for discussion)

Route the `f32 -> f16` case of `arith.truncf` through the same
`ConstrainedFPTruncIntr` used by `tt.fp_to_fp`. Sketch:

```cpp
auto inElemTy = getElementType(op.getIn());
if (inElemTy.isF32() && outElemTy.isF16()) {
  return {FpToFpOpConversion::convertFp32ToFp16(
      loc, rewriter, operands[0][0], RoundingMode::RTNE)};
}
return {LLVM::FPTruncOp::create(rewriter, loc, elemTy, operands[0][0])};
```

Only the `f32 -> f16` branch needs to change. The `bf16` branch is not
affected: it goes through `intel::convertFp32ToBf16` (either the
`__spirv_ConvertFToBF16INTEL` intrinsic or a hand-written RTNE
bit-manipulation sequence) which never emits an `fptrunc`. Confirmed
locally: `matmul(x, w).to(bf16).float()` is bit-exact against eager on
XPU today (diff = 0.0).

Happy to send a PR once the direction is agreed. Suggested tests:
- lit test in `test/TritonIntelGPU/` asserting `arith.truncf f32 -> f16`
  lowers to `llvm.experimental.constrained.fptrunc`.
- Python numerical test in `python/test/unit/` running the
  hand-written kernel above and asserting bit-exact fp16 rounding.

## Environment details

- Triton: `triton-xpu 3.7.2+git5fcc14d9`
- PyTorch (downstream): `2.14.0.dev20260713+xpu`
- GPU: Intel(R) Data Center GPU Max 1100
- Driver: `intel-opencl-icd 25.18.33578.42`, `level-zero 1.1.0`
- OS: Ubuntu 22.04.5 LTS

**Comment by anmyachev** [comment]
Hi @ZhaoqiongZ, thanks!

That sounds reasonable. We’ll be happy to review your pull request.

**Comment by ZhaoqiongZ** [comment]
Update — I built with the proposed patch (route `arith.truncf f32 -> f16` through the same `ConstrainedFPTruncIntr` used by `tt.fp_to_fp`) and it turns out **the constrained intrinsic is not enough on the XPU backend**. The fold survives.

## What I verified

Rebuilt `libtriton.so` with the patch. In the generated LLIR the barrier is now a constrained intrinsic:

```llvm
%73 = tail call half @llvm.experimental.constrained.fptrunc.f16.f32(
        float %64, metadata !"round.tonearest", metadata !"fpexcept.ignore") #6
```

so the frontend change is doing what we expected. But downstream (SPIR-V translation / IGC) still folds the `constrained.fptrunc` + `fpext` pair as identity — a hand-written round-trip kernel returns the un-rounded input:

```python
@triton.jit
def _rt(X, Y, N: tl.constexpr):
    off = tl.arange(0, N)
    x = tl.load(X + off)
    y = x.to(tl.float16).to(tl.float32)
    tl.store(Y + off, y)

x = torch.tensor([-2.643845, 1.234567, 3.14159265, -0.1000001,
                  65500.0, -1e-4, 12345.678, 0.333333],
                 device='xpu', dtype=torch.float32)
expected = x.to(torch.float16).to(torch.float32)   # CPU/eager reference
y = torch.empty_like(x); _rt[(1,)](x, y, 8)

# Observed:
#   expected: [-2.64453125, 1.234375, 3.140625, -0.0999755859..., 65504.0, ...]
#   actual  : [-2.643845...,  1.234567...,  3.1415927...,  -0.10000009..., 65500.0, ...]
#   max abs diff = 4.0  (65500 -> should be 65504)
```

The PyTorch UT `test_emulate_precision_casts_preserves_explicit_precision_cast_float16` remains failing with the same ~0.018 diff as before.

So `LLVM::ConstrainedFPTruncIntr` alone is not a hard barrier on this backend.

**Comment by ZhaoqiongZ** [comment]
Given that `ConstrainedFPTruncIntr` isn't a hard barrier on the XPU backend, resolving this correctly requires a decision on the intended lowering strategy (SPIR-V native narrowing, IGC-side fix, or something else) that only the maintainers can make. Handing this back to the team. Hi @anmyachev

**Comment by anmyachev** [comment]
Hi @ZhaoqiongZ, so I'm not sure we can fix the problem on our side and it looks like we need a fix from IGC.

We're lucky; a similar issue has already been reported to them. The fix's already publicly available; we just need to upgrade to the next rolling driver. However, it still won't work correctly on the LTS driver. Is this ok?

**Comment by ZhaoqiongZ** [comment]
Hi @anmyachev , we could mark it not working on the LTS driver, thanks for the information.

**Comment by anmyachev** [comment]
> Hi [@anmyachev](https://github.com/anmyachev) , we could mark it not working on the LTS driver, thanks for the information.

Good! I'll open this issue for now so we can track IGC fix and add a unit test to our code.

**Comment by ZhaoqiongZ** [comment]
Hi @anmyachev , I updated to latest rolling driver 26.22.38646.7 and still has this ut failure, do you know which driver version will contain the fix and when will it release officially?

**Comment by anmyachev** [comment]
> Hi [@anmyachev](https://github.com/anmyachev) , I updated to latest rolling driver 26.22.38646.7 and still has this ut failure, do you know which driver version will contain the fix and when will it release officially?

Hi @ZhaoqiongZ, I guess the minimum one is 26.27.39122.11.
